Fix webview plot not rendering after expanding

seeM opened the pull request:
Attempts to address #4088.

seeM committed:
fix webview plot not rendering after expanding

juliasilge reviewed (commented):
Ah unfortunately, with this PR and a dev build, I still see the problem:

https://github.com/user-attachments/assets/d2617ccd-7264-46c6-bb7e-b30dc2419cd0

In fact, with this PR I now can't get the plot to show up again via resizing. 😬

Are there logs you want me to check? Or please let me know if you want to pair to more fully understand why I'm seeing this.

seeM committed:
load dependencies sequentially

seeM committed:
fire size/position changed events on any resize

This is to initiate React renders during CSS animations as the pane is expanded/collapsed.

seeM commented:
I managed to repro that on my side. I think you might've encountered _another_ issue as well where htmlwidget dependencies don't always load in the correct order.

Could you please try again on the latest commit? This should also fix https://github.com/posit-dev/positron/issues/3969 🤞.

juliasilge commented:
Ah yessssssss 🚀 this is working way way better for me now.

I tried out with [highcharter](https://github.com/posit-dev/positron/issues/3969#issue-2403225405), [leaflet](https://github.com/posit-dev/positron/issues/3756#issuecomment-2198766058), and [plotly](https://github.com/posit-dev/positron/issues/4088#issue-2419277716) and everything looks very improved; I no longer observe the problem when you close and open the plots pane.

I want to highlight one remaining (new?) intermittent problem I am observing. Sometimes (maybe 1 out of 2 times) I don't see the HTML pop up the *first* time when I create the interactive graph, but it will show up when I mouse over (not click into) the plots pane. By this, I mean on a very first launch of a Positron dev build. (Possible this is a dev build only problem?) This seems to happen more often if the Positron app is sized sort of small-ish (like laptop sized) and I have it in the "stacked" layout so that the plots pane is fairly small.

juliasilge commented:
Also addresses https://github.com/posit-dev/positron/issues/3968

seeM committed:
clean up

seeM committed:
return a promise that resolves once rendering is completed

seeM commented on extensions/positron-r/resources/js/htmlwidget.js:
```diff
@@ -52,54 +52,52 @@ const arrayify = (obj) => {
  *  loaded.
  */
 const renderDependencies = (dependencies) => {
-	return new Promise((resolve, _reject) => {
```
This change is to ensure that scripts are appended _and_ loaded in sequence. I'm not sure why the browser wasn't already doing that, but this change does seem to fix highcharter and leaflet widgets.

seeM commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsView.tsx:
```diff
@@ -176,20 +179,6 @@ export class PositronPlotsViewPane extends PositronViewPane implements IReactCom
 		}));
 	}
 
-	/**
```
`this._positronReactRenderer.destroy` was deprecated in favor of the disposable interface so I snuck that in too.

seeM commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsView.tsx:
```diff
@@ -206,8 +195,31 @@ export class PositronPlotsViewPane extends PositronViewPane implements IReactCom
 		this._positronPlotsContainer = DOM.$('.positron-plots-container');
 		container.appendChild(this._positronPlotsContainer);
 
+		// Observe the plots container for resizes and fire size/position changed events.
```
This fixes the layout when expanding/collapsing the plots pane or neighboring panes.

seeM commented:
@juliasilge interesting! Do you mind if we log the new issue separately? I was unable to repro it locally trying with different window sizes and widget types.

juliasilge commented:
Sure thing! I will wait to see if I still see it in a release build, I think.

jmcphers commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsView.tsx:
```diff
@@ -206,8 +195,31 @@ export class PositronPlotsViewPane extends PositronViewPane implements IReactCom
+		// since the expand/collapse transition may be animated. Otherwise, the size/position
+		// changed events would only fire at the beginning of the animation possibly leading
+		// to incorrect layouts.
+		this._positronPlotsContainerResizeObserver = new ResizeObserver(entries => {
```
Is there a chance the observer could already exist here from a previous call to `renderBody`? If so, do we need to disconnect it before overwriting it to avoid accumulating observers?

seeM replied:
Good point, addressed in ae5dfdd.

jmcphers reviewed (dismissed):
Only reviewed the code since Julia already took a look at the behavior.

seeM committed:
clean up resize observer on re-render

juliasilge commented:
FWIW just to wrap up possible loose ends, I tried out some plotting with build 107 and I did not observe this problem at all:

> I want to highlight one remaining (new?) intermittent problem I am observing. Sometimes (maybe 1 out of 2 times) I don't see the HTML pop up the first time when I create the interactive graph, but it will show up when I mouse over (not click into) the plots pane. By this, I mean on a very first launch of a Positron dev build. (Possible this is a dev build only problem?) This seems to happen more often if the Positron app is sized sort of small-ish (like laptop sized) and I have it in the "stacked" layout so that the plots pane is fairly small.
